File: src/com/ashcraftmedia/daftman/scene/ScoreBoard.java
package com.ashcraftmedia.daftman.scene;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.FontMetrics;
import java.awt.Graphics;
import java.awt.Image;
import java.awt.Point;

import com.ashcraftmedia.daftman.core.DaftMan;
import com.ashcraftmedia.daftman.core.ImageStore;



/**
 * HW10: DAFTMAN
 * I worked on this assignment alone, using course materials, previous work,
 * the online Java API.
 *  
 * Scoreboard.java
 * This class represents a scoreboard to display relevant information during gameplay.
 * 
 * @author Ryan Ashcraft
 * @version 1.0 12/03/2010
 */

public class ScoreBoard {
	private Image smallHeartImage;
	
	private Point loc;
	private Dimension size;
	
	private int seconds;
	private int health;
	private int score;
	private int rupeesLeft;
	private int level;
	
	/**
	 * Constructor for ScoreBoard objects.
	 * 
	 * @param aLoc The origin location
	 * @param aSize The size
	 */
	public ScoreBoard(Point aLoc, Dimension aSize) {
		loc = aLoc;
		size = aSize;
		
		smallHeartImage = ImageStore.get().getImage("SMALL_HEART");
	}
	
	/**
	 * Draws strings and images to display important information during gameplay.
	 * 
	 * @param g The Graphics object
	 */
	public void draw(Graphics g) {		
		g.setColor(Color.BLACK);
		g.fillRect(loc.x, loc.y, size.width, size.height);
		
		final int OUTER_MARGIN = 16;
		FontMetrics fm = g.getFontMetrics(DaftMan.font);
		
		g.setFont(DaftMan.font);
		g.setColor(Color.WHITE);
		
		String levelString = DaftMan.addExtraSpaces(String.format("Level %02d", level));
		g.drawString(levelString, loc.x + (size.width - fm.stringWidth(levelString)) / 2, loc.y + OUTER_MARGIN*2);

		String scoreString = DaftMan.addExtraSpaces(String.format("Score %02d", score));
		g.drawString(scoreString, loc.x + OUTER_MARGIN, loc.y + OUTER_MARGIN*2);
								
		for (int i = 0; i < health; i++) {
			g.drawImage(smallHeartImage, loc.x + OUTER_MARGIN + (smallHeartImage.getWidth(null)+1)*i,
					loc.y + size.height - OUTER_MARGIN - smallHeartImage.getHeight(null),
					smallHeartImage.getWidth(null),
					smallHeartImage.getHeight(null),
					null);
		}
		
		String timeString = DaftMan.addExtraSpaces(String.format("Time %02d", seconds));
		g.drawString(timeString, size.width + loc.x*2 - OUTER_MARGIN - fm.stringWidth(timeString), loc.y + OUTER_MARGIN*2);

		String rupeesLeftString = DaftMan.addExtraSpaces(String.format("Rupees %02d", rupeesLeft));
		g.drawString(rupeesLeftString, size.width + loc.x*2 - OUTER_MARGIN - fm.stringWidth(rupeesLeftString), loc.y + OUTER_MARGIN*2 + fm.getHeight());
	}
	
	/**
	 * Sets the time.
	 * 
	 * @param aSeconds The new time (in seconds)
	 */
	public void setTime(int aSeconds) {
		seconds = aSeconds;
	}
	
	/**
	 * Sets the health.
	 * 
	 * @param aHealth The health of the bro
	 */
	public void setHealth(int aHealth) {
		health = aHealth;
	}
	
	/**
	 * Sets the score.
	 * 
	 * @param aScore The score
	 */
	public void setScore(int aScore) {
		score = aScore;
	}
	
	/**
	 * Sets the number of rupees left.
	 *  
	 * @param rupeeCount The number of rupees left
	 */
	public void setRupeesLeft(int rupeeCount) {
		rupeesLeft = rupeeCount;
	}
	
	/**
	 * Sets the level being played.
	 * 
	 * @param aLevel The level being played
	 */
	public void setLevel(int aLevel) {
		level = aLevel;
	}
}

File: web/js/scene/ScoreBoard.js
import DaftMan from '../core/DaftMan.js';
import ImageStore from '../core/ImageStore.js';

const OUTER_MARGIN = 16;

export default class ScoreBoard {
    constructor(loc, size) {
        this._loc = loc;
        this._size = size;

        this._smallHeartImage = ImageStore.get().getImage('SMALL_HEART');

        this._seconds = 0;
        this._health = 0;
        this._score = 0;
        this._rupeesLeft = 0;
        this._level = 0;
    }

    draw(ctx) {
        ctx.fillStyle = '#000';
        ctx.fillRect(this._loc.x, this._loc.y, this._size.width, this._size.height);

        ctx.font = '26px ArcadeClassic, monospace';
        ctx.fillStyle = '#fff';

        const levelString = DaftMan.addExtraSpaces(
            `Level ${String(this._level).padStart(2, '0')}`
        );
        const scoreString = DaftMan.addExtraSpaces(
            `Score ${String(this._score).padStart(2, '0')}`
        );
        const timeString = DaftMan.addExtraSpaces(
            `Time ${String(this._seconds).padStart(2, '0')}`
        );
        const rupeesLeftString = DaftMan.addExtraSpaces(
            `Rupees ${String(this._rupeesLeft).padStart(2, '0')}`
        );

        // Measure text for right-alignment
        const levelWidth = ctx.measureText(levelString).width;
        const timeWidth = ctx.measureText(timeString).width;
        const rupeesWidth = ctx.measureText(rupeesLeftString).width;
        const fontHeight = 26;

        ctx.textAlign = 'left';
        ctx.textBaseline = 'top';

        // Level centered at top
        ctx.fillText(
            levelString,
            this._loc.x + (this._size.width - levelWidth) / 2,
            this._loc.y + OUTER_MARGIN * 2
        );

        // Score left-aligned
        ctx.fillText(
            scoreString,
            this._loc.x + OUTER_MARGIN,
            this._loc.y + OUTER_MARGIN * 2
        );

        // Time right-aligned
        ctx.fillText(
            timeString,
            this._size.width + this._loc.x * 2 - OUTER_MARGIN - timeWidth,
            this._loc.y + OUTER_MARGIN * 2
        );

        // Rupees right-aligned below time
        ctx.fillText(
            rupeesLeftString,
            this._size.width + this._loc.x * 2 - OUTER_MARGIN - rupeesWidth,
            this._loc.y + OUTER_MARGIN * 2 + fontHeight
        );

        // Heart images for health (bottom-left)
        if (this._smallHeartImage) {
            const heartW = this._smallHeartImage.width;
            const heartH = this._smallHeartImage.height;
            for (let i = 0; i < this._health; i++) {
                ctx.drawImage(
                    this._smallHeartImage,
                    this._loc.x + OUTER_MARGIN + (heartW + 1) * i,
                    this._loc.y + this._size.height - OUTER_MARGIN - heartH,
                    heartW,
                    heartH
                );
            }
        }
    }

    setTime(seconds) {
        this._seconds = seconds;
    }

    setHealth(health) {
        this._health = health;
    }

    setScore(score) {
        this._score = score;
    }

    setRupeesLeft(rupeeCount) {
        this._rupeesLeft = rupeeCount;
    }

    setLevel(level) {
        this._level = level;
    }
}
